### PR TITLE
Get the initial application up and running

### PR DESCRIPTION
**Pull request by RJ722**
`tests/test-data` contains mock data for testing purposes so you might want to skip it while reviewing this PR.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -0,0 +1,39 @@
+vulture-whitelist-generators
+============================
+
+Create whitelists to tackle false positives in Vulture automatically for
+frameworks using Python bindings for C and C++ libraries, for eg. PyQt.
```
Comment (The-Compiler) [inline README.rst]: nitpick: "e.g." already means "for example", so the "for" is unneeded there: https://en.wiktionary.org/wiki/e.g.

File: setup.cfg [hunk 2]
```diff
@@ -0,0 +1,5 @@
+[tool:pytest]
+addopts = --cov vulture_whitelist --cov-report=html --cov-report=term --cov-append --cov-config .coveragerc
```
Comment (The-Compiler) [inline setup.cfg]: nitpick: I don't think you need the `--cov-config` there, `.coveragerc` should be the default.

File: README.rst [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+``pip``::
+
+    $ pip install vulture_whitelist
+
+
+From local repository::
+
+    $ git clone https://github.com/RJ722/vulture-whitelist-generators
+    $ cd vulture-whitelist-generators
+    $ pip install lxml
```
Comment (The-Compiler) [inline README.rst]: Why does the user need to install lxml separately? Shouldn't `pip install .` do that automatically because it's listed as `install_requires`?

File: setup.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+import sys
+
+from setuptools.command.test import test as TestCommand
+import setuptools
+
+from vulture_whitelist.main import __version__
+
+
+class PyTest(TestCommand):
```
Comment (The-Compiler) [inline setup.py]: The official way to do this is [using the pytest-runner project](https://docs.pytest.org/en/latest/goodpractices.html#integrating-with-setuptools-python-setup-py-test-pytest-runner) - in my experience, self-grown integrations like this come with subtle problems down the line.

That being said, using `tox` to run tests in a standardized way became kind of a common practice and using `setup.py test` is done less and less, so I'd personally just not do that at all.

File: setup.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
+#! /usr/bin/env python
+
+import sys
+
+from setuptools.command.test import test as TestCommand
+import setuptools
+
+from vulture_whitelist.main import __version__
```
Comment (The-Compiler) [inline setup.py]: Importing things from your package before it's installed can lead to issues - see [this guide](https://packaging.python.org/guides/single-sourcing-package-version/) for a number of alternatives.

File: setup.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        'Programming Language :: Python :: 3',
+        'Programming Language :: Python :: 3.4',
+        'Programming Language :: Python :: 3.5',
+        'Programming Language :: Python :: 3.6',
+        'Programming Language :: Python :: Implementation :: CPython',
+        'Programming Language :: Python :: Implementation :: PyPy',
+        'Topic :: Software Development :: Quality Assurance'
+    ],
+    install_requires=['lxml'],
+    tests_require=['lxml', 'pytest', 'pytest-cov', 'vulture'],
```
Comment (The-Compiler) [inline setup.py]: nitpick: I don't think you need to list `lxml` here because it's in `install_requires`.

File: tests/__init__.py [hunk 7]
```diff
@@ -0,0 +1,6 @@
+import os.path
+
+DIR = os.path.dirname(os.path.abspath(__file__))
+REPO = os.path.dirname(DIR)
+TESTS = os.path.join(REPO, 'tests')
```
Comment (The-Compiler) [inline tests/__init__.py]: Since you're doing this in `tests/__init__.py`, the `DIR` variable should already be the same as `TESTS`, no?

File: vulture_whitelist/main.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+def get_creator(framework):
+    return CREATORS.get(framework)
+
+
+def _parse_args():
+    version = "vulture-whitelist {0}".format(__version__)
+    parser = argparse.ArgumentParser(prog='vulture-whitelist')
+    parser.add_argument('--version', action='version', version=version)
+    parser.add_argument('framework', choices=CREATORS.keys(), help=(
+        'Framweork to generate a whitelist for.'))
```
Comment (The-Compiler) [inline vulture_whitelist/main.py]: Framweork -> Framework

File: vulture_whitelist/main.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        'Name of the whitelist.'))
+    parser.add_argument('--sip', nargs=1, default=['sip'], help=(
+        'The sip binary to use when creating a whitelist for PyQt.'))
+    return parser.parse_args()
+
+
+def main():
+    args = _parse_args()
+    creator = get_creator(args.framework)
+    creator(args).create()
```
Comment (The-Compiler) [inline vulture_whitelist/main.py]: I'm not sure whether it's a good idea to pass the entire argparse namespace to the creator. You can see that it's a questionable API because you need to create a `MockArgs` in the tests.

Instead, maybe `QtWhitelistCreator` should take separate `sip` and `whitelist` arguments, and `main.py` should take care of passing those arguments to it (because it also parses the arguments, after all)?

File: tests/test_qt_xml.py [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    c = QtWhitelistCreator(args)
+    with open(args.name, 'w') as outfile:
+        xmldata = os.path.join(TESTS, 'test-data', 'xml-data')
+        for xmldir in sorted(os.listdir(xmldata)):
+            tmp = os.path.join(xmldata, xmldir)
+            c._write_mod_whitelist(
+                outfile, xmldir, c.name_set(tmp))
+    expected_whitelist = os.path.join(SAMPLE_WHITELISTS, 'qtall.py')
+    assert filecmp.cmp(args.name, expected_whitelist, shallow=False)
+    os.remove(args.name)
```
Comment (The-Compiler) [inline tests/test_qt_xml.py]: This won't remove the file if the test fails - instead, you should probably use pytest's [tmpdir fixture](https://docs.pytest.org/en/latest/tmpdir.html) to construct an output file, then pytest will automatically take care of the cleanup, and no files in the current working directory are touched.

File: tests/test_qt_xml.py [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    args = MockArg()
+    c = QtWhitelistCreator(args)
+    with open(args.name, 'w') as outfile:
+        xmldata = os.path.join(TESTS, 'test-data', 'xml-data')
+        for xmldir in sorted(os.listdir(xmldata)):
+            tmp = os.path.join(xmldata, xmldir)
+            c._write_mod_whitelist(
+                outfile, xmldir, c.name_set(tmp))
+    expected_whitelist = os.path.join(SAMPLE_WHITELISTS, 'qtall.py')
+    assert filecmp.cmp(args.name, expected_whitelist, shallow=False)
```
Comment (The-Compiler) [inline tests/test_qt_xml.py]: Whoa! I wasn't aware of the `filecmp` module, that sounds useful! :+1:

File: tests/test_script.py [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+
+def create_whitelist_from_test_sip_files():
+    path = os.path.join(TESTS, 'test-data', 'sip')
+    subprocess.call(['vulture-whitelist', 'sip'], cwd=path)
+
+
+def test_qt_whitelist():
+    create_whitelist_from_test_sip_files()
+    whitelist_generated = os.path.join(
+        TESTS, 'test-data', 'sip', 'whitelist.py')
```
Comment (The-Compiler) [inline tests/test_script.py]: Like above, consider using `tmpdir`

File: tests/test_script.py [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    whitelist_sample = os.path.join(SAMPLE_WHITELISTS, 'qtbluetooth.py')
+    subprocess.call(['cat', whitelist_generated])
```
Comment (The-Compiler) [inline tests/test_script.py]: I don't understand what you're doing here - why call `cat` on the file?

File: tests/test_with_vulture.py [hunk 14]
```diff
@@ -0,0 +1,34 @@
+from . import SAMPLE_WHITELISTS
+
+import os
+import vulture
+
+v = vulture.Vulture(verbose=True)
+#  Feed the pre-generated whitelist to Vulture.
+v.scavenge([os.path.join(SAMPLE_WHITELISTS, 'qtall.py')])
```
Comment (The-Compiler) [inline tests/test_with_vulture.py]: Doing things like this at import time is usually a bad idea (because even if you e.g. filter out the tests for this module, the code will still run).

Instead, consider using a [pytest fixture](https://docs.pytest.org/en/latest/tmpdir.html) which returns a  prepared `Vulture` object. That way, you can also be sure you always get a new object for every test (if you add more in the future), which reduces coupling between tests.

File: tests/test_with_vulture.py [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+
+
+def check(items_or_names, expected_names):
```
Comment (The-Compiler) [inline tests/test_with_vulture.py]: It's a bit weird to have a function which takes two different types of objects and then uses `isinstance` to do something different based on the type. Maybe having a `check_names` and a `check_items` would be better?

File: tests/test_with_vulture.py [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+        assert items_or_names == set(expected_names)
+    else:
+        # items_or_names is a list of vulture.Item objects.
+        names = sorted(item.name for item in items_or_names)
+        expected_names = sorted(expected_names)
+        assert names == expected_names
+
+
+def test_closeEvent_method():
+    code = """\
```
Comment (The-Compiler) [inline tests/test_with_vulture.py]: nitpick: Using [textwrap.dedent](https://docs.python.org/3/library/textwrap.html) you can indent the block of code in the test sources.

File: vulture_whitelist/main.py [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+    'sip': qt.QtWhitelistCreator
+}
+
+
+def get_creator(framework):
+    return CREATORS.get(framework)
+
+
+def _parse_args():
+    version = "vulture-whitelist {0}".format(__version__)
```
Comment (The-Compiler) [inline vulture_whitelist/main.py]: nitpick: You can use unnumbered format arguments (`{}` instead of `{0}`) with Python 3.

File: vulture_whitelist/qt.py [hunk 18]
```diff
@@ -0,0 +1,84 @@
+import itertools
+import os
+import subprocess
+import tempfile
+
+from .utils import log
```
Comment (The-Compiler) [inline vulture_whitelist/qt.py]: nitpick: The PEP8 styleguide [recommends](https://www.python.org/dev/peps/pep-0008/#imports) always using absolute imports, i.e. `from vulture_whitelist.utils import log`.

File: vulture_whitelist/qt.py [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+
+FEATURES = ['PyQt_Accessibility', 'PyQt_SessionManager', 'PyQt_SSL',
+            'PyQt_qreal_double', 'Py_v3', 'PyQt_PrintDialog', 'PyQt_Printer',
+            'PyQt_PrintPreviewWidget', 'PyQt_PrintPreviewDialog',
+            'PyQt_RawFont', 'PyQt_OpenGL', 'PyQt_Desktop_OpenGL',
+            'PyQt_NotBootstrapped']
+PLATFORMS = ['WS_X11', 'WS_WIN', 'WS_MACX']
+TIMELINE = ['Qt_5_0_0', 'Qt_5_0_1', 'Qt_5_0_2', 'Qt_5_1_0', 'Qt_5_1_1',
+            'Qt_5_2_0', 'Qt_5_2_1', 'Qt_5_3_0', 'Qt_5_3_1', 'Qt_5_3_2',
+            'Qt_5_4_0', 'Qt_5_4_1', 'Qt_5_4_2', 'Qt_5_5_0']
```
Comment (The-Compiler) [inline vulture_whitelist/qt.py]: Those constants should probably be updated based on a recent PyQt release - not sure anymore where I found them exactly, but searching for e.g. `PyQt_NotBootstrapped` in the PyQt sources should hopefully reveal where they're defined.

File: vulture_whitelist/qt.py [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+    def __init__(self, args):
+        self.sip = args.sip[0]
+        self.whitelist_name = args.name[0]
+
+    def _write_mod_whitelist(self, f, module, name_set):
+        f.write('# {}\n'.format(module))
+        for name in sorted(name_set):
+            f.write('{}.{}\n'.format(module, name))
+        f.write('\n')
+        f.flush()
```
Comment (The-Compiler) [inline vulture_whitelist/qt.py]: Why is this needed? Whatever calls this method should take care of closing the file properly before e.g. reading it, and then this shouldn't be needed I think.

File: vulture_whitelist/utils.py [hunk 21]
```diff
@@ -0,0 +1,2 @@
+def log(msg):
```
Comment (The-Compiler) [inline vulture_whitelist/utils.py]: You might want to consider using [python's logging module](https://docs.python.org/3/library/logging.html) (also see the [basic tutorial](https://docs.pytest.org/en/latest/tmpdir.html)) instead.

File: vulture_whitelist/qt.py (hunk 20 above)
Comment (RJ722) [reply vulture_whitelist/qt.py]: Yes, absolutely - How can I forget this? 😅

File: tests/test_with_vulture.py (hunk 15 above)
Comment (RJ722) [reply tests/test_with_vulture.py]: That I stole from Vulture - https://github.com/jendrikseipp/vulture/blob/master/tests/__init__.py#L17-L25 😛

File: tests/test_with_vulture.py (hunk 14 above)
Comment (RJ722) [reply tests/test_with_vulture.py]: Yes, that was bad on my part

File: tests/test_script.py (hunk 13 above)
Comment (RJ722) [reply tests/test_script.py]: Sorry! - This is me being lazy while debugging. 😛

File: tests/test_qt_xml.py (hunk 10 above)
Comment (RJ722) [reply tests/test_qt_xml.py]: Yes, this is what we tried initially in https://github.com/jendrikseipp/vulture/pull/121#discussion_r193107285, but it turns out that a `NamedTemporaryFile` cannot be opened for a second time in Windows. :-(

File: vulture_whitelist/main.py (hunk 9 above)
Comment (RJ722) [reply vulture_whitelist/main.py]: I've tried the following solution:

Register a new creator using the `utils.Creator`. `name` is a required argument for now and everything else (which we can treat as metadata about our new creator object) sticks as an attribute.

File: tests/__init__.py (hunk 7 above)
Comment (RJ722) [reply tests/__init__.py]: Actually I was first doing it in a local scope in `tests/test_xyz/py` and copied it here blindly. Nice catch!

File: setup.py (hunk 5 above)
Comment (RJ722) [reply setup.py]: Thanks! May be we should make this change in Vulture also.

By far, I think the following is the simplest solution as we already using an `__init__.py`:
```py
import vulture_whitelist

setup(
    ...
    version=vulture_whitelist.__version__
    ...
)
```

File: setup.py (hunk 4 above)
Comment (RJ722) [reply setup.py]: I don't have much (or any) experience working with `setup.py` 😛  so I am gonna be gullible and leave it to your judgement. Removing these for now - We can add them later if someone complains.

File: README.rst (hunk 3 above)
Comment (RJ722) [reply README.rst]: Yes, you are right!

File: setup.cfg (hunk 2 above)
Comment (RJ722) [reply setup.cfg]: Agreed.

File: MANIFEST.in [hunk 22]
```diff
@@ -0,0 +1,3 @@
+include *.rst
+aasdjflkasjdfkljasf
```
[the remark this answers is not in the document]
Comment (RJ722) [reply MANIFEST.in]: Oh my God!

File: README.rst (hunk 1 above)
Comment (RJ722) [reply README.rst]: I've been doing it wrong my whole life, thanks for educating me! :-)

File: .travis.yml [hunk 23]
```diff
@@ -0,0 +1,30 @@
+language: python
+python:
+- '3.4'
```
[the remark this answers is not in the document]
Comment (RJ722) [reply .travis.yml]: I thought that since vulture supports it and we don't have any special workarounds for this version, we can support it.

File: .travis.yml [hunk 24]
```diff
@@ -0,0 +1,30 @@
...
+- '3.6'
+- 3.7-dev
+
+matrix:
+  fast_finish: true
+
+before_install:
+  - sudo mkdir -p /downloads
+  - sudo chmod a+rw /downloads
+  - curl -L https://sourceforge.net/projects/pyqt/files/sip/sip-4.19.8/sip-4.19.8.tar.gz -o /downloads/sip.tar.gz
```
[the remark this answers is not in the document]
Comment (RJ722) [reply .travis.yml]: I was also wondering that there should be a better way of doing so. I tried all the versions above and a few more permutations, but everything seems to fail :-(

Builds: https://travis-ci.org/RJ722/vulture-whitelist-generators/builds

File: vulture_whitelist/qt.py (hunk 19 above)
Comment (RJ722) [reply vulture_whitelist/qt.py]: Found them! I am gonna run the new updated version on the latest PyQt release and then gonna update test-data accordingly as soon as the build finishes!

File: .travis.yml (hunk 24 above)
[the remark this answers is not in the document]
Comment (The-Compiler) [reply .travis.yml]: Oh wow - I just double-checked, and apparently on Debian/Ubuntu, the `sip` commandline tool isn't part of the `python-sip` package, instead, it's in a separate [`sip-dev` package](https://packages.debian.org/stretch/sip-dev). So maybe you can try again with that package instead? Note that the first two variants shouldn't need `sudo: required`.

File: .travis.yml (hunk 23 above)
[the remark this answers is not in the document]
Comment (The-Compiler) [reply .travis.yml]: Fair enough! Let's just keep it in mind if it even gets a burden then.

File: setup.py (hunk 5 above)
Comment (The-Compiler) [reply setup.py]: That seems like it has the same issue - it tries to import `vulture_whitelist` in the script which is responsible for installing `vulture_whitelist`. I think this works as long as your `__init__.py` doesn't require any dependencies (directly or indirectly), but still, it's kind of risky IMHO.

Indeed it looks like vulture does the same thing - @jendrikseipp has nobody complained about that so far? :wink:

File: tests/test_qt_xml.py (hunk 10 above)
Comment (The-Compiler) [reply tests/test_qt_xml.py]: pytest's `tmpdir` fixture is different though: it doesn't open the file (unless you tell it to), but it takes care of creating a nicely named temporary directory and cleaning it up.

It creates a directory like `/tmp/pytest-of-florian/pytest-18/test_entering_and_leaving_hint0/` which also makes it possible to take a look for debugging easily - and at the end of a test, it automatically takes care of cleanup.

It works a bit different from Python's `tempfile` module, but I can highly recommend to use it anywhere you need a temporary location for tests with pytest.

File: tests/test_with_vulture.py (hunk 15 above)
Comment (The-Compiler) [reply tests/test_with_vulture.py]: well then ¯\\_(ツ)_/¯ :laughing:

File: setup.py (hunk 5 above)
Comment (jendrikseipp) [reply setup.py]: So far no complaints :-)

From the options at https://packaging.python.org/guides/single-sourcing-package-version/ I like option 1 best. So something like pip and qutebrowser are doing. I'll let you guys figure this one out for this generator repo, then we can use it for Vulture as well :-)

File: tests/test_qt_xml.py (hunk 10 above)
Comment (jendrikseipp) [reply tests/test_qt_xml.py]: Sounds good. Rahul, can you use this technique for the Vulture test as well?

Comment (RJ722) [reply tests/test_qt_xml.py]: That indeed sounds neat, I've updated the tests! Thanks!

> Rahul, can you use this technique for the Vulture test as well?

@jendrikseipp yeah, sure! :-)

File: setup.py (hunk 5 above)
Comment (RJ722) [reply setup.py]: Option 1 sounds robust. I've tried Does https://github.com/RJ722/vulture-whitelist-generators/pull/1/commits/aad70ab17bfec8e389292b3c645d0d4b6d931c3a fulfill?

File: .travis.yml (hunk 24 above)
[the remark this answers is not in the document]
Comment (RJ722) [reply .travis.yml]: `sip-dev` package still ships with `sip 4.18.1`, whereas we are running `sip 4.19`. [Tests broke! :-( ](https://travis-ci.org/RJ722/vulture-whitelist-generators/jobs/390072189)

File: setup.py (hunk 5 above)
Comment (RJ722) [reply setup.py]: I've opened https://github.com/jendrikseipp/vulture/issues/130 to keep track.

**Comment by The-Compiler** [comment]
I have some things to comment on, but the vast amount of test data makes this impossible to review :laughing: 

I still think it's probably easiest to keep the test data in the repository though - I'll merge this and open issues for the things I think you still should look at.
